src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,202 @@
+/*-------------------------------------------------------------------------
+ *
+ * local_multi_copy.c
+ *    Commands for running a copy locally
+ *
+ * For each local placement, we have a buffer. When we receive a slot
+ * from a copy, the slot will be put to the corresponding buffer based
+ * on the shard id. When the buffer size exceeds the threshold a local
+ * copy will be done. Also If we reach to the end of copy, we will send
+ * the current buffer for local copy.
+ *
+ * The existing logic from multi_copy.c and format are used, therefore
+ * even if user did not do a copy with binary format, it is possible that
+ * we are going to be using binary format internally.
+ *
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "commands/copy.h"
+#include "catalog/namespace.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+#include "nodes/makefuncs.h"
+#include "safe_lib.h"
+#include <netinet/in.h> /* for htons */
+
+#include "distributed/transmit.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/local_executor.h"
+#include "distributed/local_multi_copy.h"
+#include "distributed/shard_utils.h"
+
+static int ReadFromLocalBufferCallback(void *outbuf, int minread, int maxread);
+static void AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest,
+							CopyOutState localCopyOutState);
+
+static bool ShouldSendCopyNow(StringInfo buffer);
+static void DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId,
+						CopyStmt *copyStatement, bool isEndOfCopy);
+static bool ShouldAddBinaryHeaders(StringInfo buffer, bool isBinary);
+
+/*
+ * LocalCopyBuffer is used in copy callback to return the copied rows.
+ * The reason this is a global variable is that we cannot pass an additional
+ * argument to the copy callback.
+ */
+static StringInfo LocalCopyBuffer;
+
+/*
+ * WriteTupleToLocalShard adds the given slot and does a local copy if
+ * this is the end of copy, or the buffer size exceeds the threshold.
+ */
+void
+WriteTupleToLocalShard(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, int64
+					   shardId,
+					   CopyOutState localCopyOutState)
+{
+	/* since we are doing a local copy, the following statements should use local execution to see the changes */
+	TransactionAccessedLocalPlacement = true;
+
+	bool isBinaryCopy = localCopyOutState->binary;
+	if (ShouldAddBinaryHeaders(localCopyOutState->fe_msgbuf, isBinaryCopy))
```

Comment:
High-level observation: binary COPY should always be ok because we're not leaving the database. The problem with binary COPY is that the binary format sometimes contains OIDs which may not match on different servers.

Reply:
Are you suggesting that we should always use binary copy for local then?

I guess that would require us to keep some more state in copyOutState maybe such as columnOutputFunctions since we the current implementation uses the same format for remote and local therefore copyDest->columnOutputFunctions is used for both.

Reply:
I'm fine with leaving it as is for now.